tests/mockserver.py

```diff
@@ -190,17 +197,37 @@ def render(self, request):
 
 class MockServer():
 
+    HTTP = 'mock_http'
+    HTTPS = 'mock_https'
+
+    @classmethod
+    def from_mock(cls, path, is_secure=False):
+        key = cls.HTTP
+        if is_secure:
+            key = cls.HTTPS
+        return os.environ[key] + path
+
     def __enter__(self):
         from scrapy.utils.test import get_testenv
-        self.proc = Popen([sys.executable, '-u', '-m', 'tests.mockserver'],
+
+        http_port = str(get_ephemeral_port())
+        https_port = str(get_ephemeral_port())
+
+        self.proc = Popen([sys.executable, '-u', '-m', 'tests.mockserver', http_port, https_port],
                           stdout=PIPE, env=get_testenv())
         self.proc.stdout.readline()
+
+        self._oldenv = os.environ.copy()
+        os.environ[self.HTTP] = 'http://localhost:%s' % (http_port, )
+        os.environ[self.HTTPS] = 'https://localhost:%s' % (https_port, )
+
         return self
 
     def __exit__(self, exc_type, exc_value, traceback):
         self.proc.kill()
         self.proc.wait()
         time.sleep(0.2)
+        os.environ = self._oldenv
```

Comment:
does it work, i.e. does it replace environment of a proces, or does it just replace attribute of the os module? If that's the former, then why do we need to use env variables, instead of using global variables (class attributes?)

Reply:
copied this solution from ProxyConnectTestCase.tearDown

Reply:
It is a variable, not a module https://stackoverflow.com/a/43487762/8640733